packages/cli/src/lib/workerPool.ts

```diff
@@ -0,0 +1,150 @@
+// Direct from the Node.js docs
+// https://nodejs.org/api/async_context.html#using-asyncresource-for-a-worker-thread-pool
+
+import { AsyncResource } from 'async_hooks';
+import { EventEmitter } from 'events';
+import { Worker } from 'worker_threads';
+import assert from 'assert';
+import { warn } from 'console';
+
+const kTaskInfo = Symbol('kTaskInfo');
+const kWorkerFreedEvent = Symbol('kWorkerFreedEvent');
+
+type CallbackFn = (err: Error | null, result: any) => void;
+
+type Task = {
+  task: any;
+  callback: CallbackFn;
+};
+
+class WorkerPoolTaskInfo extends AsyncResource {
+  constructor(public callback: CallbackFn) {
+    super('WorkerPoolTaskInfo');
+  }
+
+  done(err: Error, result: any) {
+    this.runInAsyncScope(this.callback, null, err, result);
+    this.emitDestroy(); // `TaskInfo`s are used only once.
+  }
+}
+
+export const DEFAULT_ADD_WORKER_DELAY = 5000;
+
+type WorkerInfo = {
+  created: Date;
+  firstTask?: Date;
+};
+
+export default class WorkerPool extends EventEmitter {
+  workers: Worker[] = [];
+  freeWorkers: Worker[] = [];
+  tasks: Task[] = [];
+  workerInfoByThreadId: Map<number, WorkerInfo> = new Map();
+
+  addWorkerInterval?: NodeJS.Timeout | undefined;
+
+  constructor(
+    public taskFile: string | URL,
+    public numThreads: number,
+    public addWorkerDelay = DEFAULT_ADD_WORKER_DELAY
+  ) {
+    super();
+
+    this.addNewWorker();
+    if (numThreads > 1) this.addNewWorker();
+    if (numThreads > 2) {
+      this.addWorkerInterval = setInterval(() => {
+        this.addNewWorker();
+
+        if (this.addWorkerInterval && this.workers.length === numThreads)
+          clearInterval(this.addWorkerInterval);
+      }, addWorkerDelay);
```

Comment:
This should be [unref](https://nodejs.org/api/timers.html#immediateunref)'d, otherwise the process won't exit early if all jobs have finished processing early.